src/libraries/System.Text.Json/source_generation/src/JsonSourceGeneratorHelper.Generate.cs

```diff
@@ -0,0 +1,672 @@
+                : $"new {JsonPropertyInfoTypeName}[{properties.Count}]";
+
+            StringBuilder sb = new();
+
+            sb.Append($@"private static {JsonPropertyInfoTypeName}[] {propInitFuncVarName}(JsonSerializerContext context)
+        {{
+            JsonContext {JsonContextVarName} = (JsonContext)context;
+            JsonSerializerOptions options = context.Options;
+
+            {JsonPropertyInfoTypeName}[] {PropVarName} = {propertyArrayInstantiationValue};
+");
+
+            if (properties != null)
+            {
+                for (int i = 0; i < properties.Count; i++)
+                {
+                    PropertyMetadata memberMetadata = properties[i];
+
+                    TypeMetadata memberTypeMetadata = memberMetadata.TypeMetadata;
+
+                    string clrPropertyName = memberMetadata.ClrName;
+
+                    string declaringTypeCompilableName = memberMetadata.DeclaringTypeCompilableName;
+
+                    string memberTypeFriendlyName = memberTypeMetadata.ClassType == ClassType.TypeUnsupportedBySourceGen
+                        ? "null"
+                        : $"{JsonContextVarName}.{memberTypeMetadata.FriendlyName}";
+
+                    string typeTypeInfoNamedArg = $"propertyTypeInfo: {memberTypeFriendlyName}";
+
+                    string jsonPropertyNameNamedArg = memberMetadata.JsonPropertyName != null
+                        ? @$"jsonPropertyName: ""{memberMetadata.JsonPropertyName}"""
+                        : "jsonPropertyName: null";
+
+                    string getterNamedArg = memberMetadata.HasGetter
+                        ? $"getter: static (obj) => {{ return (({declaringTypeCompilableName})obj).{clrPropertyName}; }}"
+                        : "getter: null";
+
+                    string setterNamedArg;
+                    if (memberMetadata.HasSetter)
+                    {
+                        string propMutation = declaringTypeIsValueType
+                            ? @$"{{ Unsafe.Unbox<{declaringTypeCompilableName}>(obj).{clrPropertyName} = value; }}"
+                            : $@"{{ (({declaringTypeCompilableName})obj).{clrPropertyName} = value; }}";
+
+                        setterNamedArg = $"setter: static (obj, value) => {propMutation}";
+                    }
+                    else
+                    {
+                        setterNamedArg = "setter: null";
+                    }
+
+                    JsonIgnoreCondition? ignoreCondition = memberMetadata.IgnoreCondition;
+                    string ignoreConditionNamedArg = ignoreCondition.HasValue
+                        ? $"ignoreCondition: JsonIgnoreCondition.{ignoreCondition.Value}"
+                        : "ignoreCondition: default";
+
+                    string encodedNameArg;
+                    if (!ContainsNonAscii(clrPropertyName))
+                    {
+                        byte[] name = Encoding.UTF8.GetBytes(memberMetadata.JsonPropertyName ?? clrPropertyName);
```

Comment:
Did we decide using the default encoder is sufficient for property names? i.e. instead we would do this at run-time using the encoder specified on `JsonSerializerOptions`.

Reply:
> i.e. instead we would do this at run-time using the encoder specified on JsonSerializerOptions.

That could be a custom encoder, right? Is the idea to potentially ignore it and use the default one? Would using the default be faster or equivalent to handwriting the encoded `byte[]` as I'm doing now?

Reply:
https://github.com/dotnet/runtime/pull/51300#discussion_r614481118